src/query/server/query.go

```diff
@@ -694,23 +707,56 @@ func Run(runOpts RunOptions) RunResult {
 }
 
 // make connections to the m3db cluster(s) and generate sessions for those clusters along with the storage
-func newM3DBStorage(
-	cfg config.Configuration,
-	clusters m3.Clusters,
-	poolWrapper *pools.PoolWrapper,
-	runOpts RunOptions,
-	queryContextOptions models.QueryContextOptions,
-	tsdbOpts m3.Options,
-	downsamplerReadyCh chan<- struct{},
-	clusterNamespacesWatcher m3.ClusterNamespacesWatcher,
-	rwOpts xio.Options,
-	clockOpts clock.Options,
-	instrumentOptions instrument.Options,
-) (storage.Storage, clusterclient.Client, downsample.Downsampler, cleanupFn, error) {
+func newM3DBStorage(cfg config.Configuration, clusters m3.Clusters, poolWrapper *pools.PoolWrapper,
+	queryContextOptions models.QueryContextOptions, tsdbOpts m3.Options,
+	clusterNamespacesWatcher m3.ClusterNamespacesWatcher, instrumentOptions instrument.Options,
+) (storage.Storage, cleanupFn, *etcdclient.Configuration, error) {
+	fanoutStorage, storageCleanup, err := newStorages(clusters, cfg,
+		poolWrapper, queryContextOptions, tsdbOpts, instrumentOptions)
+	if err != nil {
+		return nil, nil, nil, errors.Wrap(err, "unable to set up storages")
+	}
+	logger := instrumentOptions.Logger()
+	cleanup := func() error {
+		lastErr := storageCleanup()
+		// Don't want to quit on the first error since the full cleanup is important
+		if lastErr != nil {
+			logger.Error("error during storage cleanup", zap.Error(lastErr))
+		}
+
+		clusterNamespacesWatcher.Close()
+
+		if err := clusters.Close(); err != nil {
+			lastErr = errors.Wrap(err, "unable to close M3DB cluster sessions")
+			// Make sure the previous error is at least logged
+			logger.Error("error during cluster cleanup", zap.Error(err))
+		}
+
+		return lastErr
+	}
+
+	etcdConfig := cfg.ClusterManagement.Etcd
+	if etcdConfig == nil && len(cfg.Clusters) == 1 &&
+		cfg.Clusters[0].Client.EnvironmentConfig != nil {
+		syncCfg, err := cfg.Clusters[0].Client.EnvironmentConfig.Services.SyncCluster()
+		if err != nil {
+			return nil, nil, nil, errors.Wrap(err, "unable to get etcd sync cluster config")
+		}
+		etcdConfig = syncCfg.Service
+	}
```

Comment:
Unless I'm missing something, there is nothing that would require having this etcd code inside `newM3DBStorage()` function. Could it be moved to where it was before, or extracted to a separate function?

Reply:
It seems to be coupled to m3db use case since it depends on Clusters setup. I can extract it to a separate function and move its call to a `switch case` part for m3db.

Reply:
> I can extract it to a separate function and move its call to a switch case part for m3db.

👍